src/msg/async/EventEpoll.cc

```diff
@@ -23,16 +26,90 @@
 #undef dout_prefix
 #define dout_prefix *_dout << "EpollDriver."
 
+#define BUILD_BUG_ON(condition) ((void )sizeof(char [1 - 2*!!(condition)]))
+#define READ_ONCE(v) (*(volatile decltype(v)*)&(v))
+
+#ifdef __x86_64__
+#ifndef __NR_sys_epoll_create2
+#define __NR_sys_epoll_create2  428
+#endif
+#endif
+
+static inline long epoll_create2(int flags, size_t size)
+{
+#ifdef __NR_sys_epoll_create2
+  return syscall(__NR_sys_epoll_create2, flags, size);
+#else
+  errno = ENOSYS;
+  return -1;
+#endif
+}
+
+int EpollDriver::uepoll_mmap(int epfd,
+			     struct epoll_uheader **_header,
+			     unsigned int **_index)
+{
+  struct epoll_uheader *header;
+  unsigned int *index, len;
+  int rc;
+
+  len = sysconf(_SC_PAGESIZE);
+again:
+  header = (decltype(header))mmap(NULL, len, PROT_WRITE|PROT_READ,
+				  MAP_SHARED, epfd, 0);
+  if (header == MAP_FAILED) {
+    rc = -errno;
+    lderr(cct) << "mmap(header): " << cpp_strerror(errno) << dendl;
+    return rc;
+  }
+
+  if (header->header_length != len) {
+    unsigned int tmp_len = len;
+
+    len = header->header_length;
+    munmap(header, tmp_len);
+    goto again;
+  }
+  ceph_assert(header->magic == EPOLL_USERPOLL_HEADER_MAGIC);
+
+  index = (decltype(index))mmap(NULL, header->index_length,
+				PROT_WRITE|PROT_READ, MAP_SHARED,
+				epfd, header->header_length);
+  if (index == MAP_FAILED) {
+    rc = -errno;
+    lderr(cct) << "mmap(index): " << cpp_strerror(errno) << dendl;
+    return rc;
+  }
+
+  BUILD_BUG_ON(sizeof(*header) != EPOLL_USERPOLL_HEADER_SIZE);
+  BUILD_BUG_ON(sizeof(header->items[0]) != 16);
+
+  *_header = header;
+  *_index = index;
+
+  return 0;
+}
+
 int EpollDriver::init(EventCenter *c, int nevent)
 {
+  use_uepoll = g_ceph_context->_conf.get_val<bool>("ms_uepoll");
   events = (struct epoll_event*)malloc(sizeof(struct epoll_event)*nevent);
   if (!events) {
     lderr(cct) << __func__ << " unable to malloc memory. " << dendl;
     return -ENOMEM;
   }
   memset(events, 0, sizeof(struct epoll_event)*nevent);
 
-  epfd = epoll_create(1024); /* 1024 is just an hint for the kernel */
+  if (use_uepoll) {
+    epfd = epoll_create2(EPOLL_USERPOLL, 1024);
+    if (epfd >= 0) {
+      /* Mmap all pointers */
+      uepoll_mmap(epfd, &uheader, &uindex);
+      lderr(cct) << "UEPOLL enabled!" << dendl;
+    }
+  } else {
+    epfd = epoll_create1(0);
```

Comment:
can we take this chance to pass `FD_CLOEXEC` to it?

Reply:
Sure, but who invokes exec() ? I mean is there somewhere in ceph where fork() and exec() are called? Because otherwise I do not understand why it is needed.